osu.Desktop/Windows/WindowsAssociationManager.cs

```diff
@@ -0,0 +1,273 @@
+
+                NotifyShellUpdate();
+            }
+            catch (Exception e)
+            {
+                Logger.Error(e, @"Failed to uninstall file and URI associations.");
+            }
+        }
+
+        public static void NotifyShellUpdate() => SHChangeNotify(EventId.SHCNE_ASSOCCHANGED, Flags.SHCNF_IDLIST, IntPtr.Zero, IntPtr.Zero);
+
+        /// <summary>
+        /// Installs or updates associations.
+        /// </summary>
+        private static void updateAssociations()
+        {
+            using var classes = Registry.CurrentUser.OpenSubKey(SOFTWARE_CLASSES, true);
+            Debug.Assert(classes != null);
+
+            foreach (var association in file_associations)
+                association.Install(classes, EXE_PATH);
+
+            foreach (var association in uri_associations)
+                association.Install(classes, EXE_PATH);
+        }
+
+        private static void updateDescriptions(LocalisationManager? localisation)
+        {
+            using var classes = Registry.CurrentUser.OpenSubKey(SOFTWARE_CLASSES, true);
+            Debug.Assert(classes != null);
+
+            foreach (var association in file_associations)
+                association.UpdateDescription(classes, getLocalisedString(association.Description));
+
+            foreach (var association in uri_associations)
+                association.UpdateDescription(classes, getLocalisedString(association.Description));
+
+            string getLocalisedString(LocalisableString s)
+            {
+                if (localisation == null)
+                    return s.ToString();
+
+                var b = localisation.GetLocalisedBindableString(s);
+                b.UnbindAll();
+                return b.Value;
+            }
+        }
+
+        #region Native interop
+
+        [DllImport("Shell32.dll")]
+        private static extern void SHChangeNotify(EventId wEventId, Flags uFlags, IntPtr dwItem1, IntPtr dwItem2);
+
+        private enum EventId
+        {
+            /// <summary>
+            /// A file type association has changed. <see cref="Flags.SHCNF_IDLIST"/> must be specified in the uFlags parameter.
+            /// dwItem1 and dwItem2 are not used and must be <see cref="IntPtr.Zero"/>. This event should also be sent for registered protocols.
+            /// </summary>
+            SHCNE_ASSOCCHANGED = 0x08000000
+        }
+
+        private enum Flags : uint
+        {
+            SHCNF_IDLIST = 0x0000
+        }
+
+        #endregion
+
+        private record FileAssociation(string Extension, LocalisableString Description, string IconPath)
+        {
+            private string programId => $@"{PROGRAM_ID_PREFIX}{Extension}";
+
+            /// <summary>
+            /// Installs a file extenstion association in accordance with https://learn.microsoft.com/en-us/windows/win32/com/-progid--key
+            /// </summary>
+            public void Install(RegistryKey classes, string exePath)
```

Comment:
Is there a point to having all these method accept `SOFTWARE_CLASSES` but opened as the first param? Can that just be inlined here?

Reply:
It's to avoid repeatedly opening and closing the same registry key.

Reply:
Is that an issue though? I'm looking at this from a maintainability angle, wherein it's a public method anyone can point a random registry key at. I'd like our interactions with the registry to be very particular and limited in scope.